### DISCUSSION TITLE
data_array

### DISCUSSION THREAD

**Discussion by giscus** [ruanqizhen/labview_book#31] [Announcements]
# 数组和循环结构

https://lv.qizhen.xyz/data_array

**Comment by Arita9527** [comment]
以前老版本有一个字符串查询并且输出的控件，我可以从一个字符串中取出某个（段）字符。
新版本却不见了，只能数组的查询。如下图。
![](https://i.postimg.cc/90h14MY8/searchIP.png)

问题：
去哪里能够找到以前的字符串查询控件呢？
或者是新版本如何达到这个目的呢。谢谢

**Reply by ruanqizhen** [reply]
不记得老版本是啥样了，如果只是查找一段子字符串，可以就用上文里提到的"Match Pattern“（"匹配模式" ）函数试一下：
输入一个长字符串，比如”abcdef“，匹配模式是个子字符串，比如”cd“。

**Comment by alvin-lrl** [comment]
hi，Mr. Ruan.
留意到文中提到的数组为空的时候需要特别处理否则运算会出错，该如何解决呢？
![](https://raw.githubusercontent.com/alvin-lrl/LabVIEW_code/main/test1114.png)
从图中的运行结果来看，确实为空的最后一行是没有结果输出的。

**Reply by ruanqizhen** [reply]
这是 LabVIEW 的默认行为，两个不同长度的数组运算，结果数组长度由原本两个数组中较短的那个决定。如果你需要不同的结果，比如结果的长度要跟原来较长的那个数组一样，就只能自己进行处理：先比较两个数组长度，然后按照某种规则把短数组的数据补长，让两个数组一样长再运算。

**Reply by alvin-lrl** [reply]
Thanks！

**Comment by alvin-lrl** [comment]
来交作业了
https://github.com/alvin-lrl/LabVIEW_code/blob/main/practise_1114.png

**Reply by XuY6969** [reply]
download图片时，显示无法访问此网站，github.com 的响应时间过长。

**Reply by alvin-lrl** [reply]
我这边可以正常打开，检查一下网络叭

**Comment by alvin-lrl** [comment]
a new question came up
https://github.com/alvin-lrl/LabVIEW_code/blob/main/test1114.png

**Reply by ruanqizhen** [reply]
这个问题看一下输出隧道这里：https://lv.qizhen.xyz/data_array#%E8%BE%93%E5%87%BA%E9%9A%A7%E9%81%93

**Reply by alvin-lrl** [reply]
天呐，直接拖出来改成indexing就可以了，我也不知道上午为啥不能弄出来，感谢！

**Comment by XuY6969** [comment]
用校园网和自己手机热点都打不开&nbsp;


990352597
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人: ***@***.***&gt;; 
发送时间: 2022年12月10日(星期六) 晚上9:40
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [ruanqizhen/labview_book] data_array (Discussion #31)





 
我这边可以正常打开，检查一下网络叭

**Comment by Riching5** [comment]
关于显示sin的强度图的问题：
想请问一下那个强度图的x坐标是怎么显示0~300，我设置横坐标最大值最小值为300和0，不起作用，用常量赋给属性节点的x起始值和最大值也不行。是不是这个x坐标本身显示的是系统的时间，我要怎么才能使x坐标代表输入二维数组的行数。谢谢您，祝您小年快乐。

**Reply by ruanqizhen** [reply]
我估计你用的是“强度图表”（Intensity Chart）控件，所以无法修改 x 范围。“强度图”（Intensity Graph）控件是很容易设置 x 坐标的。这两个控件类似，容易搞混。[图形化显示数据](https://lv.qizhen.xyz/data_graph)一节会介绍这两种控件的区别。

**Reply by ruanqizhen** [reply]
“强度图表”（Intensity Chart）控件也可以设置 x 坐标长度，但是要先改图表历史长度（Chart History Length），再改 x 坐标最大最小范围。

**Comment by Riching5** [comment]
确实是用混了，感谢。
&nbsp;


雷秦
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人: ***@***.***&gt;; 
发送时间: 2023年1月14日(星期六) 晚上10:21
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [ruanqizhen/labview_book] data_array (Discussion #31)





 
我估计你用的是“强度图表”（Intensity Chart）控件，所以无法修改 x 范围。“强度图”（Intensity Graph）控件是很容易设置 x 坐标的。这两个控件类似，容易搞混。图形化显示数据一节会介绍这两种控件的区别。

**Comment by Riching5** [comment]
感谢，两种方式都成功实现
&nbsp;


雷秦
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人: ***@***.***&gt;; 
发送时间: 2023年1月14日(星期六) 晚上10:26
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [ruanqizhen/labview_book] data_array (Discussion #31)





 
“强度图表”（Intensity Chart）控件也可以设置 x 坐标长度，但是要先改图表历史长度（Chart History Length），再改 x 坐标最大最小范围。

**Comment by Riching5** [comment]
在做for循环第一个例子时摸索出来几点经验分享一下：注意运行前路径输入控件不能为空，不然会出现报错“错误8发生于 Recursive File List.vi-”；因为是想打开一个文件夹，所以可以右键路径输入控件属性，将浏览选项-选择模式-文件夹勾选；进度条的属性标尺的最大值改为1，不然进度条会看起来没有动。

**Comment by Riching5** [comment]
想请问最后一个例子，在我高亮执行过程后就出现13和245的结果，我也尝试在其中一个循环里加了一个1s延时后结果变成了15（加延时）和234（没加）。是不是两个for循环因为并行，所以最终的结果会受哪个循环跑的快，跑的快的一边就会更频繁使用子VI里移位寄存器的结果？

**Reply by ruanqizhen** [reply]
是的，可以这么理解。事例中，可以确定的是两个循环的总迭代次数：上面一个两次，下面一个三次。但哪个先执行，哪个后执行是不确定的，与循环里其它的代码也有关。

**Comment by XXWJ4Zzzz** [comment]
您好，在节“簇数据运算”的第二段话末尾“还可以前后两个簇中元素顺序不一致等引发的错误”处存在语法错误，应在“可以”后面添加“避免”、“防止”一类的词汇。

**Comment by XXWJ4Zzzz** [comment]
在“数组”的“输出通道”第三条“Concatening”处，文尾的“每一次循环迭代传递给隧道会被附加的数组的后面”表意不清，应修改为类似“每一次循环迭代传递给隧道会被添加在对应输出内容的后面”的表述。

**Reply by ruanqizhen** [reply]
多谢指正！刚刚把这两处错误修改了。

**Comment by XXWJ4Zzzz** [comment]
您好，在节“移位寄存器”中的“改进后的程序”字样下的第二段文尾“除非能够确定程序的不需要给移位寄存器赋初值”应该为“除非能够确定程序的确不需要给移位寄存器赋初值”。

**Comment by ruanqizhen** [comment]
之前博客上的相关文章，搬到这里：

### 循环运行某个时间后退出

有时候需要一个循环不是在迭代多少次之后停下，而是在运行多少时间之后停下。一般，直接的解决方案就是利用“时间计数器”函数，进入循环前，记下当前时间，然后每次循环迭代都查看一下当前时间，若超过所需时间，则退出循环。这个功能也可以使用“已用时间”Express VI来完成。

把“已用时间”Express VI拖到程序框图上，会出现它的配置对话框。我们需要这个Express VI计时5秒，每次计时完成自动重置。

在循环里，需要用到的是“已用时间”的“结束”输出，一旦运行时间超过5秒，这个值就会被设为真。这时循环停止。

但是下图程序有个问题，就是“耗时40毫秒”这个完成功能的VI和“已用时间”VI之间没有连线，它们是同时运行的，而“已用时间”运行时间几乎可以忽略，一调用它，它就立刻返回“结束”值。这时，即便“结束”值为“真”，循环也要再等大约40毫秒，待功能完成功能不部分运行结束，循环才停止。

强制程序先工作，再计时间是否可以解决这个问题呢，如下图中的程序。这样也不行，因为“已用时间”是在重置后，第一次调用它的时候开始计时的。下图这个程序，循环第一次迭代，并没有马上就开始计时，而是要等到功能VI完成后，已经耗用的一段时间，才开始计时。这个即时已经不精确了。

所以要完全解决这个问题，只能麻烦一点。程序中多加一个条件语句。在循环一开始，就立刻计时；而在后续的迭代中，每次功能完成再检查当前时间。或者，把“已用时间”放在前面，每次判断是否结束。如果是，则不执行“耗时40毫秒”这个VI。

**Reply by ruanqizhen** [reply]
之前博客下的留言：

Qizhen
2008年09月12日 10:40 编辑
“耗时40毫秒”这个VI只是用来演示的，所以也没有加error in/out第一和第三种方案在于第一种方案要多延时一个迭代周期，在这里就是40毫秒。比如一个极端的设置，我们设定0秒后循环停止。正确的行为就应当是“耗时40毫秒”这个VI一次都不执行。第三种方案可以做到，但是第一种就不行，它至少也要运行“耗时40毫秒”VI一次。当然，这几种方案一般都用在定时精度要求不高的应用场合，任务VI被多执行一次问题也不大，为了程序简介，用第一种方案也未尝不可。

回复
傻傻的
2008年09月12日 09:37 编辑
其实感觉第3中办法和第1种办法也没什么区别啊。一样是要等耗时40毫秒的子VI结束了才能结束。比如说设置的已用时间不是耗时的子VI的所用时间的整数倍的话。这两种方法就没什么区别了。反而感觉第一中方法来得干净利落，呵呵。个人见解，请大侠指点

回复
杰
2008年09月9日 19:46 编辑
耗时40ms.vi怎么没有error in and out呢？
没有错误处理的子VI不符合LV的编程规则，虽然程序没有问题，呵呵。
把这两个VI的错误簇输入并联到一起，大概就差不多了吧。不过搞不清到底LV会先执行哪个，所谓的 race condition 是吧？
但是本来，用软件来定时就不精确，虽然ms量级的效果还凑合。

回复
亮
2008年09月9日 18:14 编辑
其实可以直接把时间判定做在序列第一格,然后拉一条线从下面(不透过第二个序列)把结束与否传给循环终止就好.

回复
WL
2008年09月9日 17:47 编辑
学习

回复
五月的红枫
2008年09月9日 13:53 编辑
学习了．．．

回复
五月的红枫
2008年09月9日 13:53 编辑
沙发！

回复

**Reply by ruanqizhen** [reply]
之前博客下的留言：

jahnson
2012年06月23日 13:56 编辑
本博文中第五个VI，延迟3次后我实验的结果为2，而不是3了，请阮大侠再验证一下了；但是第六个VI的结果是9了，跟我验证的相符了，如果这样那后面3*3=9的解释好像就不正确了，没法传图上来，谅解。端午佳节快乐！

回复
ruanqizhen
2012年06月25日 09:37 编辑
这个地方我写错了，之后有人发现了错误，但我还一直没来得及改正。既然又被指出了，我今天就直接在原文上更新一下。

回复
ruanqizhen
2012年06月25日 09:45 编辑
改好了

回复
静
2011年05月31日 17:37 编辑
请教一下和这个主题不大相关的问题。可以讲一下关于用labview编写半峰宽或者峰面积方面的知识吗？ 半峰宽有两种计算方法：
一种是半高宽法fwhm，即做峰底的切线L，在峰高一半的地方做L平行线。一种是积分法，做峰底的切线L，测量峰的面积，测量峰的高度，
用“面积/高度”得到峰宽。我采用第二种方法，自己编写了一个VI，采用labview自带的积分模块，但是，上下限的动态确定有点问题，峰面积求取结果不对。您认为有什么方法解决吗？非常感谢

回复
ruanqizhen
2011年06月2日 08:59 编辑
我对信号处理方面没有了解，这个问题帮不了你了。

回复
Pingback： 《我和LabVIEW》补充内容和注释 « 凡人琐忆 编辑
ruanqizhen
2011年05月27日 13:06 编辑
删除数组需要重新排列数组元素位置，所以效率低。
LabVIEW的最初的目标是针对非专业程序员得，所以并没有透露很多细节给用户。现在LabVIEW这么庞大，可能没人弄得清所有底层的东西了。对于一些特定的问题，可能主要还是得靠编程者自己摸索。或者去论坛询问一下，集众人之力。

回复
huangchen
2011年05月27日 14:53 编辑
唉，个人认为如果底层用的链表的话，重排位置应该也不会太耗时间啊呵呵可能使用的那种连续的线性表吧~

回复
ruanqizhen
2011年05月27日 17:19 编辑
这个毕竟叫做“数组”，底层肯定也是真正的数组，否则索引就很慢了。

回复
huangchen
2011年06月2日 14:23 编辑
对，我2了，您说的很有道理，谢谢~

回复
huangchen
2011年05月27日 00:02 编辑
阮大侠，您好，我使用labview也有1年多了，最近做一个很小的功能的时候，发现拆分一维数组和删除数组元素这两个函数都可以完成一个一维数组的循环拆分（使用移位寄存器），但是第一个函数要比第二个函数效率高几十倍。

我想问一下，为什么拆分一维数组要比删除数组元素效率高这么多？我这里有测试vi，不过不知道怎么发给您；
第二，个人有时候感觉labview封装的挺好，用起来非常方便，但是却不能让我们比较容易的了解到它的底层，它的某些函数如何实现的，也就不能很方便地选取更适合的函数来开发程序（有时候选用不好就会导致程序效率很低下），您认为有解决这个问题的好办法么，或者有相应的书籍材料么？

回复

**Comment by Automan-wfq** [comment]
在输入隧道那一环节中，博主写的是“在调试程序的过程中，有候会出现这种情况，明明提供了 N 的值，循环却没有迭代或迭代次数不对”，”有侯“应该是少了一个字，应该改为“在调试程序的过程中，有时候会出现这种情况，明明提供了 N 的值，循环却没有迭代或迭代次数不对”。

**Reply by ruanqizhen** [reply]
把错字修改了，多谢指正！
